java/core/src/test/java/com/lancedb/lance/DatasetTest.java

```diff
@@ -234,6 +237,61 @@ void testDropColumns() {
     }
   }
 
+  @Test
+  void testAlterColumns() {
+    String testMethodName = new Object() {}.getClass().getEnclosingMethod().getName();
+    String datasetPath = tempDir.resolve(testMethodName).toString();
+    try (RootAllocator allocator = new RootAllocator(Long.MAX_VALUE)) {
+      TestUtils.SimpleTestDataset testDataset =
+          new TestUtils.SimpleTestDataset(allocator, datasetPath);
+      dataset = testDataset.createEmptyDataset();
+      assertEquals(testDataset.getSchema(), dataset.getSchema());
+      ColumnAlteration nameColumnAlteration = new ColumnAlteration("name");
+      nameColumnAlteration.setRename(Optional.of("new_name"));
+      nameColumnAlteration.setNullable(Optional.of(Boolean.TRUE));
+      nameColumnAlteration.setDataType(Optional.of(new ArrowType.Utf8()));
+      dataset.alterColumns(Collections.singletonList(nameColumnAlteration));
+
+      Schema changedSchema =
+          new Schema(
+              Arrays.asList(
+                  Field.nullable("id", new ArrowType.Int(32, true)),
+                  Field.notNullable("new_name", new ArrowType.Utf8())),
+              null);
+
+      assertEquals(changedSchema.getFields().size(), dataset.getSchema().getFields().size());
+      assertEquals(
+          changedSchema.getFields().stream().map(Field::getName).collect(Collectors.toList()),
+          dataset.getSchema().getFields().stream()
+              .map(Field::getName)
+              .collect(Collectors.toList()));
+
+      nameColumnAlteration = new ColumnAlteration();
+      nameColumnAlteration.setPath("new_name");
+      nameColumnAlteration.setRename(Optional.of("new_name_2"));
+      nameColumnAlteration.setDataType(Optional.of(new ArrowType.LargeUtf8()));
+
+      dataset.alterColumns(Collections.singletonList(nameColumnAlteration));
```

Comment:
Have you considered a builder here? This seems much nicer IMO

```suggestion
      nameColumnAlteration = ColumnAlteration.Builder("new_name")
          .rename("new_name_2")
          .castTo(new ArrowType.LargeUtf8())
          .build();

      dataset.alterColumns(Collections.singletonList(nameColumnAlteration));
```

Reply:
@wjones127 Please take a look, have refactored.